include/ghex/transport_layer/ucx/communicator.hpp

```diff
@@ -163,25 +160,23 @@ namespace gridtools {
                         gridtools::ghex::tl::cb::progress_status status;
                         int p = 0;
                         p+= ucp_worker_progress(m_ucp_sw);
-                        p+= ucp_worker_progress(m_ucp_sw);
-                        p+= ucp_worker_progress(m_ucp_sw);
+
+                        /* this is really important for large-scale multithreading */
+                        sched_yield();
+
                         status.m_num_sends = std::exchange(m_send_worker->m_progressed_sends, 0);
-                        m_send_worker->m_thread_primitives->critical(
-                            [this,&p,&status]()
-                            {
-                                p+= ucp_worker_progress(m_ucp_rw);
-                                p+= ucp_worker_progress(m_ucp_rw);
-                                status.m_num_recvs = std::exchange(m_recv_worker->m_progressed_recvs, 0);
-                                status.m_num_cancels = std::exchange(m_recv_worker->m_progressed_cancels, 0);
-                            }
-                        );
-                        //return p;
+                        {
+                        std::lock_guard<decltype(m_send_worker->mutex())> lock(m_send_worker->mutex());
```

Comment:
here again for clarity: use the mutex from context directly (or from recv worker). remove_reference_t?

Reply:
I do not think remove_reference helps here, since this is an internal function and the fact that it works without adding facilities tells me we do not need it. Adding it, would mean, to me, that I do not understand what's going on (which may be true, btw). We decided to not make this change, since there are some issues with forward declarations otherwise, which would make the code harder to read.